src/simple/multiplex/client.rs

```diff
@@ -28,20 +28,30 @@ pub trait ClientProto<T: 'static>: 'static {
     /// Response messages.
     type Response: 'static;
 
+    /// The type of request ids to used to correlate requests to responses
+    type RequestId: RequestId;
+
     /// The message transport, which usually take `T` as a parameter.
     ///
     /// An easy way to build a transport is to use `tokio_core::io::Framed`
     /// together with a `Codec`; in that case, the transport type is
     /// `Framed<T, YourCodec>`. See the crate docs for an example.
     type Transport: 'static +
-        Stream<Item = (RequestId, Self::Response), Error = io::Error> +
-        Sink<SinkItem = (RequestId, Self::Request), SinkError = io::Error>;
+        Stream<Item = (Self::RequestId, Self::Response), Error = io::Error> +
+        Sink<SinkItem = (Self::RequestId, Self::Request), SinkError = io::Error>;
 
     /// A future for initializing a transport from an I/O object.
     ///
     /// In simple cases, `Result<Self::Transport, Self::Error>` often suffices.
     type BindTransport: IntoFuture<Item = Self::Transport, Error = io::Error>;
 
+    /// The `RequestIdSource` to use.
+    type RequestIds: RequestIdSource<Self::RequestId, Self::Request>;
```

Comment:
I think that `RequestIds` could have a better name altogether, but haven't figured out a better one yet. One possibility is just using `RequestIdSource` which I tried to initially avoid as I thought it would be confusing, but seems to work with `RequestId` assoc type matching the trait name.